vegafusion-python/src/lib.rs

```diff
@@ -505,6 +408,105 @@ impl PyVegaFusionRuntime {
         })
     }
 
+    #[allow(clippy::too_many_arguments)]
+    #[pyo3(signature = (spec, local_tz, default_input_tz=None, preserve_interactivity=None, inline_datasets=None, keep_signals=None, keep_datasets=None))]
+    pub fn pre_transform_logical_plan(
+        &self,
+        py: Python,
+        spec: PyObject,
+        local_tz: String,
+        default_input_tz: Option<String>,
+        preserve_interactivity: Option<bool>,
+        inline_datasets: Option<&Bound<PyDict>>,
+        keep_signals: Option<Vec<(String, Vec<u32>)>>,
+        keep_datasets: Option<Vec<(String, Vec<u32>)>>,
+    ) -> PyResult<(PyObject, PyObject, PyObject)> {
+        let spec = parse_json_spec(spec)?;
+        let preserve_interactivity = preserve_interactivity.unwrap_or(true);
+
+        let inline_datasets = process_inline_datasets(inline_datasets)?;
+
+        let mut keep_variables: Vec<PreTransformVariable> = Vec::new();
+        for (name, scope) in keep_signals.unwrap_or_default() {
+            keep_variables.push(PreTransformVariable {
+                variable: Some(Variable::new_signal(&name)),
+                scope,
+            });
+        }
+        for (name, scope) in keep_datasets.unwrap_or_default() {
+            keep_variables.push(PreTransformVariable {
+                variable: Some(Variable::new_data(&name)),
+                scope,
+            });
+        }
+
+        let (client_spec, export_updates, warnings) = py.allow_threads(|| {
+            self.tokio_runtime
+                .block_on(self.runtime.pre_transform_logical_plan(
+                    &spec,
+                    inline_datasets,
+                    &PreTransformLogicalPlanOpts {
+                        local_tz,
+                        default_input_tz,
+                        preserve_interactivity,
+                        keep_variables,
+                    },
+                ))
+        })?;
+
+        Python::with_gil(|py| -> PyResult<(PyObject, PyObject, PyObject)> {
+            let py_spec = pythonize::pythonize(py, &client_spec)?;
+
+            let py_export_list = PyList::empty(py);
+            for export_update in export_updates {
+                let py_export_dict = PyDict::new(py);
+                py_export_dict.set_item(
+                    "namespace",
+                    pythonize::pythonize(py, &export_update.namespace)?,
+                )?;
+                py_export_dict.set_item("name", export_update.name)?;
+
+                match export_update.value {
+                    TaskValue::Plan(plan) => {
+                        // TODO: we probably want more flexible serialization format than pg_json, but protobuf
+                        // fails with our memtable (possibly fixed in DataFusion 49)
```

Comment:
this comment seems to be incorrect because rust is using datafusion 49.0.0 (as defined in Cargo.toml line 104)